Source/Particles/Deposition/CurrentDeposition.H

```diff
@@ -126,208 +362,179 @@ void doDepositionShapeN (const GetParticlePosition<PIdx>& GetPosition,
+            const amrex::Real vx  = uxp[ip]*gaminv;
+            const amrex::Real vy  = uyp[ip]*gaminv;
+            const amrex::Real vz  = uzp[ip]*gaminv;
+
             amrex::Real wq  = q*wp[ip];
             if (do_ionization){
                 wq *= ion_lev[ip];
             }
 
-            amrex::ParticleReal xp, yp, zp;
-            GetPosition(ip, xp, yp, zp);
+            doDepositionShapeNKernel<depos_order>(xp, yp, zp, wq, vx, vy, vz, jx_arr, jy_arr, jz_arr,
+                                                  jx_type, jy_type, jz_type,
+                                                  relative_time,
+                                                  AMREX_D_DECL(dzi, dxi, dyi),
+                                                  AMREX_D_DECL(zmin, xmin, ymin),
+                                                  invvol, lo, n_rz_azimuthal_modes);
 
-            const amrex::Real vx  = uxp[ip]*gaminv;
-            const amrex::Real vy  = uyp[ip]*gaminv;
-            const amrex::Real vz  = uzp[ip]*gaminv;
-            // wqx, wqy wqz are particle current in each direction
-#if defined(WARPX_DIM_RZ)
-            // In RZ, wqx is actually wqr, and wqy is wqtheta
-            // Convert to cylindrical at the mid point
-            const amrex::Real xpmid = xp + relative_time*vx;
-            const amrex::Real ypmid = yp + relative_time*vy;
-            const amrex::Real rpmid = std::sqrt(xpmid*xpmid + ypmid*ypmid);
-            amrex::Real costheta;
-            amrex::Real sintheta;
-            if (rpmid > 0._rt) {
-                costheta = xpmid/rpmid;
-                sintheta = ypmid/rpmid;
-            } else {
-                costheta = 1._rt;
-                sintheta = 0._rt;
-            }
-            const Complex xy0 = Complex{costheta, sintheta};
-            const amrex::Real wqx = wq*invvol*(+vx*costheta + vy*sintheta);
-            const amrex::Real wqy = wq*invvol*(-vx*sintheta + vy*costheta);
-#else
-            const amrex::Real wqx = wq*invvol*vx;
-            const amrex::Real wqy = wq*invvol*vy;
+        }
+    );
+#if defined(WARPX_USE_GPUCLOCK)
+    if( load_balance_costs_update_algo == LoadBalanceCostsUpdateAlgo::GpuClock) {
+        amrex::Gpu::streamSynchronize();
+        *cost += *cost_real;
+        amrex::The_Managed_Arena()->free(cost_real);
+    }
 #endif
-            const amrex::Real wqz = wq*invvol*vz;
+}
 
-            // --- Compute shape factors
-            Compute_shape_factor< depos_order > const compute_shape_factor;
-#if (AMREX_SPACEDIM >= 2)
-            // x direction
-            // Get particle position after 1/2 push back in position
-#if defined(WARPX_DIM_RZ)
-            // Keep these double to avoid bug in single precision
-            const double xmid = (rpmid - xmin)*dxi;
-#else
-            const double xmid = ((xp - xmin) + relative_time*vx)*dxi;
+/**
+ * \brief Direct current deposition for thread thread_num for the implicit scheme
```

Comment:
Yes, templating can be done. Though with `push_type = 0` the `uxp_n` etc would be null pointers since those particle attributes would not have been defined.